### ISSUE TITLE
About Performance

### ISSUE THREAD

**Issue by ComputerMath** [TRI-ML/KP2D#12]
Hi.
Thanks for opening the code.

I trained the model as README and just increased batch size from 8(default) to 32. 
I used 2 GPUs using dataparallel to use multi-GPU and it took 2 days to complete 50 epoches to train.

The pretrained model you gave got the same performance as the paper, 
but when i trained the model following default config, the performance got even worse than the pretrained model you offered.
(As seen in below image)

![image](https://user-images.githubusercontent.com/29303833/104050438-fb688580-5229-11eb-9a67-9a02dd38c0ac.png)

Is there any different environment, config or additional dataset from the default setting in configs/base_config.py to make same as pretrained model you offered?
(e.g. the default train dataset repeat (default is 1), loss weight (default is all 1)

**Comment by RaresAmbrus** [comment]
Hi,

I did notice some variance when training after updating to pytorch 1.6, but I manged to get close to the original numbers we got in the paper with the configs in the repo, running on 8GPUs in the docker environment. Did you also use the docker environment to train? And if time is an issue you could try the v2 model, it should train about 30% faster.

**Comment by ComputerMath** [comment]
Thanks for answering.
I tried many parameter adjusting in base_config.
I changed (reduced) the batch size at the cost of time and I noticed that it would affect the performance.
For my personal opinion, because of the network not being that deep layer, the batch size would effect performance.
I could find the performance close to the performance in the paper. :)

**Comment by RaresAmbrus** [comment]
Ok great! You're right about the network being lightweight, and in our follow-up work - [Self-Supervised 3D Keypoint Learning for Ego-Motion Estimation](https://arxiv.org/pdf/1912.03426.pdf) we noticed a significant performance improvement by simply switching to a ResNet backbone (Table 4).

**Comment by ComputerMath** [comment]
I already read about the paper (KP3D). 
It was very interesting because I also have interest in depth estimation, and I'm looking forward to see the code for KP3D soon (in the github page, the code writer said it would open soon).
Thanks for your all paper.

By the way, was there any experiment that the cell size is not 8 but 4 or 16 for the KP2D?
I modified the code you offered  into somehow slower than original (about 2 times slower)
and I'm curious about the performance about that.

If there had been experiments to make various architecture about the model, can you tell me briefly how it was?

Thanks :)

**Comment by ComputerMath** [comment]
To talk about KP3D, can you tell me how much slower does is KP3D than KP2D?
I'm now entering into graduate student and I'm planning to study about sub-pixel level keypoint detector/descriptor like the model you suggested (what a inspirational papers in my study). 
What I'm thinking is this: To make the model practical enough, how much light the network should be?
Of course the faster, the better but I'm thinking that the KP3D's ResNet backbone would be somehow burden because the keypoint detector/descriptor is just one of the Large pipeline (ex: localization -> Image Retrieval + keypoint detect + PnP or pose estimation....)

**Comment by RaresAmbrus** [comment]
Thanks - it's great to hear that you're looking into this :-) For the KP2D model, we kept the cell size fixed to 8 so we can easily compare with other methods without changing the architecture. If you manage to run some experiments with other cell sizes I'm curious to see how it works! 

For speed, I think one thing that can definitely be improved is the augmentations - right now we do that per image, on the CPU, but most of those operations can be done on the GPU which would speed things up quite a bit. Also, our KP3D resnet model definitely makes training slower, but we managed to get realtime inference performance on 640x192 images by first converting the model using TorchScript. 

I'm looking into releasing the KP3D codebase, but that's still going to take some time. In the meantime I could try to port our ResNet-based keypoint network and release it in this codebase if that would be helpful?

**Comment by ComputerMath** [comment]
I understood.
I'm hoping to see the KP3D code base soon.
I will very appreciate if you can open the ResNet-base model in advance so that I can experiment the performance.
(of course i will show you performance)
Thanks for telling me in detail.

**Comment by captainfffsama** [comment]
It seems that image resolution is more import than batchsize or other fact.
I just use coco test dataset which only 5004 image, batchsize set 1.train image shape set 544*960,the final result is:
```
Train [ E 49, T 49, R 0.5632, R_Avg 0.5556, L 1.1009, L_Avg 1.3105]: 100%|████████████████████████████████| 5004/5004 [08:26<00:00,  9.88 images/s]
Loaded 580 image pairs 
Evaluating for (320, 240) -- top_k 300
evaluate_keypoint_net: 580it [00:17, 33.27it/s]
Repeatability 0.674
Localization Error 0.916
Correctness d1 0.533
Correctness d3 0.833
Correctness d5 0.898
MScore 0.548
```
maybe just use more upsample module like FPN,use a deeper backbone,get a better result is not  hard?

**Comment by RaresAmbrus** [comment]
Sorry about the slow reply. Yes, I agree, a deeper network would help, and you can also increase the top_k parameter, as when dealing with higher resolutions you will be able to get more features. I would also say that playing around a bit with the augmentations and tuning the learning rate and other parameters should help. 

In the meantime I added the KeypointResnet model we used for KP3D: https://github.com/TRI-ML/KP2D/pull/15 -- I haven't had time to tune any of the parameters or weights, and I got these numbers while testing to see if it works (I'm using an 8GPU machine and the docker configuration from the repo):
```
Loaded 580 image pairs
Evaluating for (320, 256) -- top_k 300
evaluate_keypoint_net: 580it [00:26, 21.57it/s]
Repeatability 0.668
Localization Error 0.859
Correctness d1 0.598
Correctness d3 0.878
Correctness d5 0.922
MScore 0.548
```
With a bit of tuning the localization error should go below 0.8, and correctness 1 well above 0.6. I hope this helps!

**Comment by ComputerMath** [comment]
Thanks for your work. 
The code you committed seems merge fail, so i think i have to wait some more. 
Now I can experiment many things for KP3D model thanks to you. 

The results you gave was far better than i thought.
Thanks again :)

**Comment by RaresAmbrus** [comment]
We'll have to wait for a review before merging the PR, but it should be done in the next couple of days. Good luck with the experiments :)

**Comment by RaresAmbrus** [comment]
The PR has been merged: https://github.com/TRI-ML/KP2D/commit/77633af57d65c5a0cd7609eaeab7e6cb758ca791

Closing this issue for now, let me know how things go!
